### DISCUSSION TITLE
OpenThread Border Router NAT64 issue

### DISCUSSION THREAD

**Discussion by marcorussi** [openthread/openthread#8806] [Q&A] [answered]
Hi all,

a while ago I managed to successfully build a Thread border router for my Thread experiments based on a Raspberry Pi. It is working fine as my Thread node (nrf52840 dev kit) can still reach a CoAp server in the internet (on its IPv6 address). The IPv6 address was made with the well-known NAT64 prefix. 

Now I am trying to create a new border router for a friend of mine and I am struggling to reproduce my same scenario. This new border router is connected via Ethernet to a different WiFi router in my friend's place. Another node (nrf52840 dev kit), running the same firmware of my node, can connected to the Thread network and successfully ping 8.8.8.8 from OT CLI. So everything looks good until the node tries to reach the same CoAp server in the internet. Indeed it cannot reach an IPv6 address in the internet.

Both Thread networks have NAT64 route 64:ff9b::::/96 (set via netdata register). I have checked with ot-ctl netdata show. 

I have checked the ip route (commands "route -A inet6" and "ip a") of both border routers and it is clear that the new border router does not have a NAT64 route. I do not remember if I did something in particular with my border router and if I have to do it on the new one as well. Indeed I followed the official guide from google. 
The main difference I see is a NAT64 interface and related routes as below:

ip a
```
nat64: <POINTOPOINT,MULTICAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 500
    link/none 
    inet 192.168.255.1/32 scope global nat64
       valid_lft forever preferred_lft forever
    inet6 fdaa:bb:1::2/128 scope global 
       valid_lft forever preferred_lft forever
    inet6 fe80::782d:3732:13a2:e3f3/64 scope link stable-privacy 
       valid_lft forever preferred_lft forever
```
route -A inet6
```
64:ff9b::/96                   [::]                       U    1024 5     0 nat64
fdaa:bb:1::2/128               [::]                       U    256 2     0 nat64
fe80::/64                      [::]                       U    256 1     0 nat64
fdaa:bb:1::2/128               [::]                       Un   0   3     0 nat64
fe80::782d:3732:13a2:e3f3/128  [::]                       Un   0   3     0 nat64
ff00::/8                       [::]                       U    256 1     0 nat64
```

I hope it is clear and sorry to not be more precise but I am not an expert.
Thank you.

Marco

**Comment by jwhui** [comment, marked as answer]
The latest Thread Specification is not using the well-known prefix (`64:ff9b::/96`) and, as a result, the OpenThread Border Router implementation has been updated to match.

The [Thread Border Router - Provide Internet access via NAT64](https://openthread.io/codelabs/openthread-border-router-nat64) codelab provides an overview of using NAT64.

**Reply by marcorussi** [reply]
Hi @jwhui , thank you for your answer. The nat64 feature (sudo ot-ctl nat64 enable) of OTBR was apparently not enough as I couldn't still reach an external IPv6 address (my CoAp server). I noticed that my old OTBR was running an old commit while I made the new one from latest main branch. I found that OTBR was using tayga service for bridging IPv4 and IPv6. According to what you said this service has been removed in the recent OTBR commits. Correct?
Anyway I had to manually configure and launch tayga to fix my issue and make the new OTBR reach the external IPv6 server. 
Does it make sense?
Thanks.

**Reply by jwhui** [reply]
> The nat64 feature (sudo ot-ctl nat64 enable) of OTBR was apparently not enough as I couldn't still reach an external IPv6 address (my CoAp server).

NAT64 is for reaching global IPv4 endpoints, not IPv6 endpoints.

> I found that OTBR was using tayga service for bridging IPv4 and IPv6. According to what you said this service has been removed in the recent OTBR commits. Correct?

Correct, we no longer use TAYGA. See:
- https://github.com/openthread/ot-br-posix/pull/1539

> Anyway I had to manually configure and launch tayga to fix my issue and make the new OTBR reach the external IPv6 server.
Does it make sense?

You should not need configure/install TAYGA when using the latest OTBR implementation.

**Reply by marcorussi** [reply]
Thank you for clarifying, Do you mean that I should use the NAT64 prefix provided by OTBR when enable NAT64?
For example I know get 
**fdc7:7369:247e:2:0:0::/96 sn low f400**

**Reply by jwhui** [reply]
Yes, you should use the locally provided NAT64 prefix.

The [`otNat64SynthesizeIp6Address()`](https://openthread.io/reference/group/api-nat64#otnat64synthesizeip6address) API will synthesize an IPv6 address from an IPv4 address using the advertised NAT64 prefix.

**Reply by marcorussi** [reply]
Marvellous! Thank you very much.
